Make processPythonExecution return a future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* **#57855 Make processPythonExecution return a future**
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

lw committed:
Make processPythonExecution return a future

We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]

lw committed:
Update on "Make processPythonExecution return a future"


We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]

lw committed:
Update on "Make processPythonExecution return a future"


We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]

mrshenli commented:
Test failures are real:

```
May 17 22:43:05   test_py_rref_args (__main__.ProcessGroupRpcTestWithSpawn) ... [E request_callback_no_python.cpp:620] Received error while processing request type 260: *ownerRRef->type() == *typeINTERNAL ASSERT FAILED at "/var/lib/jenkins/workspace/torch/csrc/distributed/rpc/rref_context.cpp":362, please report a bug to PyTorch. OwnerRRef type is Tensor, expected type is Tensor
May 17 22:43:05 Exception raised from getOrCreateOwnerRRef at /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/rref_context.cpp:362 (most recent call first):
May 17 22:43:05 frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x69 (0x7f4df0f01799 in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
May 17 22:43:05 frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xd2 (0x7f4df0efd8d2 in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
May 17 22:43:05 frame #2: c10::detail::torchInternalAssertFail(char const*, char const*, unsigned int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x4e (0x7f4df0eff22e in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
May 17 22:43:05 frame #3: torch::distributed::rpc::RRefContext::getOrCreateOwnerRRef(torch::distributed::rpc::GloballyUniqueId const&, std::shared_ptr<c10::Type> const&) + 0x3c1 (0x7f4de9aca2a1 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
May 17 22:43:05 frame #4: torch::distributed::rpc::RequestCallbackNoPython::processBaseScriptRemoteCall(torch::distributed::rpc::RpcCommandBase&, std::function<void (torch::distributed::rpc::Message)> const&, c10::intrusive_ptr<c10::ivalue::Future, c10::detail::intrusive_target_default_null_type<c10::ivalue::Future> > const&) const + 0x694 (0x7f4de9ab9434 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
May 17 22:43:05 frame #5: torch::distributed::rpc::RequestCallbackNoPython::processRpc(torch::distributed::rpc::RpcCommandBase&, torch::distributed::rpc::MessageType const&, std::shared_ptr<torch::distributed::rpc::LazyStreamContext>) const + 0x2a0 (0x7f4de9abf010 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
May 17 22:43:05 frame #6: torch::distributed::rpc::RequestCallbackImpl::processRpcWithErrors(torch::distributed::rpc::RpcCommandBase&, torch::distributed::rpc::MessageType const&, std::shared_ptr<torch::distributed::rpc::LazyStreamContext>) const + 0x54 (0x7f4df226d734 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
May 17 22:43:05 frame #7: <unknown function> + 0x3f9b95e (0x7f4de9abb95e in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
May 17 22:43:05 frame #8: torch::distributed::rpc::RequestCallbackNoPython::processMessage(torch::distributed::rpc::Message&, std::shared_ptr<torch::distributed::rpc::LazyStreamContext>) const + 0xb60 (0x7f4de9abccb0 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
May 17 22:43:05 frame #9: torch::distributed::rpc::RequestCallback::operator()(torch::distributed::rpc::Message&, std::shared_ptr<torch::distributed::rpc::LazyStreamContext>) const + 0x47 (0x7f4de9ab57d7 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
May 17 22:43:05 frame #10: torch::distributed::rpc::ProcessGroupAgent::handleRecv(torch::distributed::rpc::RecvWork&) + 0x1e0 (0x7f4df224b8b0 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
May 17 22:43:05 frame #11: <unknown function> + 0xd10c42 (0x7f4df224cc42 in /opt/conda/lib/python3.6/site-packages/torch/lib/libtorch_python.so)
May 17 22:43:05 frame #12: c10::ThreadPool::main_loop(unsigned long) + 0x2a3 (0x7f4df0ed7113 in /opt/conda/lib/python3.6/site-packages/torch/lib/libc10.so)
May 17 22:43:05 frame #13: <unknown function> + 0xc8421 (0x7f4de5a74421 in /opt/conda/lib/libstdc++.so.6)
May 17 22:43:05 frame #14: <unknown function> + 0x76ba (0x7f4e0033f6ba in /lib/x86_64-linux-gnu/libpthread.so.0)
May 17 22:43:05 frame #15: clone + 0x6d (0x7f4e0007551d in /lib/x86_64-linux-gnu/libc.so.6)
```

lw commented:
The test failures turned out to come from the previous PR.

lw committed:
Update on "Make processPythonExecution return a future"


We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM

lw committed:
Update on "Make processPythonExecution return a future"


We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]

lw committed:
Update on "Make processPythonExecution return a future"


We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]

lw committed:
Update on "Make processPythonExecution return a future"


We already had a helper to run Python functions, which was nice (it de-duplicated some code). This helper was however taking a callback which, as I said, isn't as nice as it returning a Future. Hence here I change this.

Differential Revision: [D28253846](https://our.internmc.facebook.com/intern/diff/D28253846/)

[ghstack-poisoned]
